File: backend/types/recurring_events.py
"""
Shared types for recurring events functionality in the backend
Maintains consistency with TypeScript types in shared-types/
"""

from typing import Dict, List, Optional, Union
from dataclasses import dataclass
from datetime import datetime


@dataclass
class RecurringConfig:
    """Configuration for recurring events"""
    active: bool
    frequency: str  # 'daily', 'weekly', 'monthly'
    days_of_week: Optional[List[int]] = None  # 0 = Sunday, 1 = Monday, etc.
    end_date: Optional[str] = None  # ISO string when to stop generating
    max_occurrences: Optional[int] = None  # Maximum number of occurrences
    timezone: Optional[str] = None  # Timezone for the recurring events


@dataclass
class RecurringEventTemplate:
    """Base event template for recurring events"""
    id: str
    club_id: str
    title: str
    caption: Optional[str] = None
    poster_url: Optional[str] = None
    ticket_link: Optional[str] = None
    start_date: str
    end_date: str
    music_genres: Optional[List[str]] = None
    created_by: str
    created_at: str
    inserted_at: str
    updated_at: str
    recurring_config: RecurringConfig


@dataclass
class GeneratedEventInstance:
    """Individual event instance generated from template"""
    club_id: str
    title: str
    caption: Optional[str] = None
    poster_url: Optional[str] = None
    ticket_link: Optional[str] = None
    start_date: str
    end_date: str
    music_genres: Optional[List[str]] = None
    created_by: str
    # Note: Individual instances don't have recurring_config


@dataclass
class RecurringEventGenerationOptions:
    """Options for generating recurring events"""
    weeks_ahead: int = 4
    dry_run: bool = False
    club_id: Optional[str] = None
    template_id: Optional[str] = None


@dataclass
class RecurringEventGenerationResult:
    """Result of recurring event generation"""
    message: str
    count: int
    generated_events: List[GeneratedEventInstance]
    warnings: Optional[List[str]] = None


def dict_to_recurring_config(data: Dict) -> RecurringConfig:
    """Convert dictionary to RecurringConfig object"""
    return RecurringConfig(
        active=data.get('active', True),
        frequency=data.get('frequency', 'weekly'),
        days_of_week=data.get('days_of_week'),
        end_date=data.get('end_date'),
        max_occurrences=data.get('max_occurrences'),
        timezone=data.get('timezone')
    )


def recurring_config_to_dict(config: RecurringConfig) -> Dict:
    """Convert RecurringConfig object to dictionary"""
    return {
        'active': config.active,
        'frequency': config.frequency,
        'days_of_week': config.days_of_week,
        'end_date': config.end_date,
        'max_occurrences': config.max_occurrences,
        'timezone': config.timezone
    }

File: shared-types/recurring-events.ts
/**
 * Shared TypeScript types for recurring events functionality
 * Used across frontend, backend, and club-dashboard
 */

export interface RecurringConfig {
  /** Whether the recurring event is active */
  active: boolean;

  /** How often the event repeats */
  frequency: "daily" | "weekly" | "monthly";

  /** Days of the week for weekly events (0 = Sunday, 1 = Monday, etc.) */
  days_of_week?: number[];

  /** When to stop generating recurring events (ISO string) */
  end_date?: string;

  /** Maximum number of occurrences to generate */
  max_occurrences?: number;

  /** Timezone for the recurring events (optional, defaults to club timezone) */
  timezone?: string;
}

export interface RecurringEventTemplate {
  /** The base event template */
  id: string;
  club_id: string;
  title: string;
  caption?: string;
  poster_url?: string;
  ticket_link?: string;
  start_date: string;
  end_date: string;
  music_genres?: string[];
  created_by: string;
  created_at: string;
  inserted_at: string;
  updated_at: string;
  recurring_config: RecurringConfig;
}

export interface GeneratedEventInstance {
  /** Individual event instance generated from template */
  club_id: string;
  title: string;
  caption?: string;
  poster_url?: string;
  ticket_link?: string;
  start_date: string;
  end_date: string;
  music_genres?: string[];
  created_by: string;
  /** Note: Individual instances don't have recurring_config */
}

export interface RecurringEventGenerationOptions {
  /** Number of weeks ahead to generate events */
  weeks_ahead?: number;

  /** Whether this is a dry run (no actual creation) */
  dry_run?: boolean;

  /** Specific club ID to generate for (optional) */
  club_id?: string;

  /** Specific event template ID to generate for (optional) */
  template_id?: string;
}

export interface RecurringEventGenerationResult {
  /** Success message */
  message: string;

  /** Number of events generated */
  count: number;

  /** Array of generated events */
  generated_events: GeneratedEventInstance[];

  /** Any warnings or additional info */
  warnings?: string[];
}
